### PR TITLE
rustdoc: overhaul code block lexing errors

### LINKED ISSUE #53919: doc: Errors during syntax highlighting appear to be fatal
When building documentation on bindgen-produced Rust files, a possible sight looks like this:

<pre><font color="#EF2929"><b>error</b></font><font color="#FFFFFF"><b>: unknown start of token: \</b></font>
 <font color="#739FCF"><b>--&gt; </b></font>&lt;stdin&gt;:8:1
  <font color="#739FCF"><b>|</b></font>
<font color="#739FCF"><b>8</b></font> <font color="#739FCF"><b>| </b></font>\__________pkt-&gt;size___________/          \_result-&gt;size_/ \__pkt-&gt;size__/                                                
  <font color="#739FCF"><b>| </b></font><font color="#EF2929"><b>^</b></font>

<font color="#FCE94F"><b>warning</b></font><font color="#FFFFFF"><b>: Backing out of syntax highlighting</b></font>
  <font color="#739FCF"><b>|</b></font>
  <font color="#739FCF"><b>= </b></font><font color="#FFFFFF"><b>note</b></font>: You probably did not intend to render this as a rust code-block</pre>

(Apparently github swallows the output color codes, but envision error in alerting red and warning in yellow).

To the moderately experienced developer, this screams "The build did not succeed", and they might even go as far as to look at how to make rust-bindgen produce more readable comments -- where actually, everything is fine and this is just a warning.

It appears to me that the error output of the highlighter is passed on to stderr. I suggest that the error be visually styled in a way that it can not be mistaken for a fatal error in the build process itself, eg. by replacing the red "error" text with a yellow "syntax highlighter error (caught)", and/or by indenting the output of the syntax highlighter.

I'd be happy to contribute code but failed to find the component that issues the "Backing out of syntax highlighting" messages; pointers would be appreciated.

## Meta
* cargo 1.29.0-nightly (0ec7281b9 2018-08-20)

### PR DESCRIPTION
**Pull request by euclio**
Fixes #53919.

This PR moves the reporting of code block lexing errors from rendering time to an early pass, so we can use the compiler's error reporting mechanisms. This dramatically improves the diagnostics in this situation: we now de-emphasize the lexing errors as a note under a warning that has a span and suggestion instead of just emitting errors at the top level.

Additionally, this PR generalizes the markdown -> source span calculation function, which should allow other rustdoc warnings to use better spans in the future.

Last, the PR makes sure that the code block is always emitted in the docs, even if it fails to highlight correctly.

Of note:
- The new pass unfortunately adds another pass over the docs to gather the doc blocks for syntax-checking. I wonder if this could be combined with the pass that looks for testable blocks? I'm not familiar with that code, so I don't know how feasible that is.
- `pulldown_cmark` doesn't make it easy to find the spans of the code blocks, so the code that calculates the spans is a little nasty. It works for all the test cases I threw at it, but I wouldn't be surprised if an edge case would break it. Should have a thorough review.
- This PR worsens the state of #56885, since those certain fatal lexing errors are now emitted before docs get generated at all.

### CODE REVIEW

File: src/test/rustdoc/bad-codeblock-syntax.rs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+/// ```
+pub fn quux() {}
+
+// @has bad_codeblock_syntax/fn.ok.html
+// @has - '//*[@class="docblock"]/pre/code[@class="language-text"]' '\_'
+/// ```text
+/// \_
+/// ```
+pub fn ok() {}
+
```
Comment (GuillaumeGomez) [inline src/test/rustdoc/bad-codeblock-syntax.rs]: nit: Extra empty line.

**Comment by GuillaumeGomez** [comment]
> The new pass unfortunately adds another pass over the docs to gather the doc blocks for syntax-checking. I wonder if this could be combined with the pass that looks for testable blocks? I'm not familiar with that code, so I don't know how feasible that is.

We could do everything in just one pass, not sure we'd like that. For instance, we check intra links and if private items have doc tests in one pass. But maybe that'd be a nice thing to only have one pass and iterate over passes for each item. To be determined...

**Comment by QuietMisdreavus** [comment]
Reading through now. This PR will probably take some time to give a real review for, though...

> The new pass unfortunately adds another pass over the docs to gather the doc blocks for syntax-checking. I wonder if this could be combined with the pass that looks for testable blocks? I'm not familiar with that code, so I don't know how feasible that is.

I wonder if we could make a kind of "item pass" so we could combine all the ones that need to check things, but *not* add new items, into one actual "pass" that runs multiple functions over each item as it finds it? That way we reduce the number of crate scans that need to happen, but still make the checks somewhat modular.

**Comment by euclio** [comment]
@QuietMisdreavus I'm unclear on the scope of the changes that you'd like me to make here. Would it suffice to just attempt to combine the passes for gathering doctests and checking their syntax?

**Comment by QuietMisdreavus** [comment]
Sorry, i was speaking hypothetically when i mentioned "item passes". This PR can stay using a separate pass to perform syntax-checking; the refactoring can come later. I'm reading over the PR now...

**Review by QuietMisdreavus — COMMENTED** [review]
One last nit, then this is ready to go!

**Comment by QuietMisdreavus** [comment]
Thanks so much! This is ready to go!

@bors r+

**Comment by QuietMisdreavus** [comment]
```
[01:55:30] ---- doc::output_not_captured stdout ----
[01:55:30] running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/cargo doc`
[01:55:30] thread 'doc::output_not_captured' panicked at '
[01:55:30] Expected: execs
[01:55:30]     but: expected to find:
[01:55:30] 1 | â˜ƒ
[01:55:30] 
[01:55:30] did not find in output:
[01:55:30]     Checking a v0.0.1 (/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/cit/t534/foo/a)
[01:55:30]  Documenting a v0.0.1 (/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/cit/t534/foo/a)
[01:55:30] warning: could not parse code block as Rust code
[01:55:30]  --> a/src/lib.rs:2:17
[01:55:30]   |
[01:55:30] 2 |               /// ```
[01:55:30]   |  _________________^
[01:55:30] 3 | |             /// â˜ƒ
[01:55:30] 4 | |             /// ```
[01:55:30]   | |___________________^
[01:55:30]   |
[01:55:30]   = note: error from rustc: unknown start of token: \u{2603}
[01:55:30] help: mark blocks that do not contain Rust code as text
[01:55:30]   |
[01:55:30] 2 |             /// ```text
[01:55:30]   |                 ^^^^^^^
[01:55:30] 
[01:55:30]  Documenting foo v0.0.1 (/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/cit/t534/foo)
[01:55:30]     Finished dev [unoptimized + debuginfo] target(s) in 4.58s
[01:55:30] ', src/tools/cargo/tests/testsuite/support/mod.rs:773:13
[01:55:30] note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
[01:55:30] 
[01:55:30] 
[01:55:30] failures:
[01:55:30]     doc::output_not_captured
```

This looks like [one of Cargo's tests](https://github.com/rust-lang/cargo/blob/2b4a5f1f0bb6e13759e88ea9512527b0beba154f/tests/testsuite/doc.rs#L653-L685), where it's scanning the output of rustdoc to make sure Cargo hasn't swallowed it. However, i'm not totally sure what the error is saying here.

@rust-lang/cargo, this PR changes the output format of doctest failures, and it seems to have tripped up one of Cargo's tests. How should we change the Cargo test to make it work? (And how should we coordinate the submodule update with this PR? `>_>`)

**Comment by dwijnand** [comment]
@QuietMisdreavus already fixed in https://github.com/rust-lang/cargo/pull/6559.  You'll have to rebase on top of #57721 to unblock this.

**Comment by euclio** [comment]
Yeah, I noticed that failure yesterday. My fix is already in cargo master, we just have to wait for #57721  to land.

**Comment by QuietMisdreavus** [comment]
Oh! Well, i'm glad it's almost already taken care of! :sweat_smile:

**Comment by Centril** [comment]
@euclio https://github.com/rust-lang/rust/pull/57721 has landed.

**Comment by euclio** [comment]
This just needs a retry, then!

**Comment by QuietMisdreavus** [comment]
On the assumption that this will Just Work...

@bors retry

**Pull request merged**